File: array/problem47/problem47.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        // Sort to group duplicates; avoid picking same value twice at the same depth
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> path;
        vector<bool> used(nums.size(), false);
        dfs(nums, used, path, res);
        return res;
    }

private:
    void dfs(const vector<int>& nums, vector<bool>& used, vector<int>& path,
             vector<vector<int>>& res) {
        if (path.size() == nums.size()) {
            res.push_back(path);
            return;
        }
        for (size_t i = 0; i < nums.size(); ++i) {
            if (used[i]) {
                continue;
            }
            if (i > 0 && !used[i - 1] && nums[i] == nums[i - 1]) {
                // Same value as previous which is not used in this position -> duplicate permutation
                continue;
            }
            used[i] = true;
            path.push_back(nums[i]);
            dfs(nums, used, path, res);
            path.pop_back();
            used[i] = false;
        }
    }
};

File: array/problem47/problem47.py
from typing import List


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        # Sort so duplicates are adjacent; skip same value at same recursion depth
        nums.sort()
        n = len(nums)
        used = [False] * n
        res: List[List[int]] = []
        path: List[int] = []

        def dfs() -> None:
            if len(path) == n:
                res.append(path[:])
                return
            prev = None
            for i in range(n):
                if used[i]:
                    continue
                if prev is not None and nums[i] == prev:
                    continue
                prev = nums[i]
                used[i] = True
                path.append(nums[i])
                dfs()
                path.pop()
                used[i] = False

        dfs()
        return res
